File: src/utils/logger.ts
export type LogLevel = 'silent' | 'error' | 'warn' | 'info' | 'debug' | 'trace';

const LOG_LEVEL_PRIORITY: Record<LogLevel, number> = {
  silent: 0,
  error: 1,
  warn: 2,
  info: 3,
  debug: 4,
  trace: 5,
};

export interface Logger {
  debug(message: string, ...meta: unknown[]): void;
  info(message: string, ...meta: unknown[]): void;
  warn(message: string, ...meta: unknown[]): void;
  error(message: string, ...meta: unknown[]): void;
  trace(message: string, ...meta: unknown[]): void;
  child(prefix: string): Logger;
}

export class DebugLogger implements Logger {
  private level: LogLevel;
  private prefix: string;
  private priority: number;

  constructor(level: LogLevel | boolean = 'warn', prefix: string = '') {
    this.prefix = prefix;
    if (typeof level === 'boolean') {
      this.level = level ? 'debug' : 'warn';
    } else {
      this.level = level;
    }
    this.priority = LOG_LEVEL_PRIORITY[this.level];
  }

  debug(message: string, ...meta: unknown[]): void {
    if (this.priority >= LOG_LEVEL_PRIORITY.debug) {
      console.debug(`[DEBUG${this.prefix ? `:${this.prefix}` : ''}] ${message}`, ...meta);
    }
  }

  info(message: string, ...meta: unknown[]): void {
    if (this.priority >= LOG_LEVEL_PRIORITY.info) {
      console.info(`[INFO${this.prefix ? `:${this.prefix}` : ''}] ${message}`, ...meta);
    }
  }

  warn(message: string, ...meta: unknown[]): void {
    if (this.priority >= LOG_LEVEL_PRIORITY.warn) {
      console.warn(`[WARN${this.prefix ? `:${this.prefix}` : ''}] ${message}`, ...meta);
    }
  }

  error(message: string, ...meta: unknown[]): void {
    if (this.priority >= LOG_LEVEL_PRIORITY.error) {
      console.error(`[ERROR${this.prefix ? `:${this.prefix}` : ''}] ${message}`, ...meta);
    }
  }

  trace(message: string, ...meta: unknown[]): void {
    if (this.priority >= LOG_LEVEL_PRIORITY.trace) {
      console.trace(`[TRACE${this.prefix ? `:${this.prefix}` : ''}] ${message}`, ...meta);
    }
  }

  child(prefix: string): Logger {
    return new DebugLogger(this.level, this.prefix ? `${this.prefix}:${prefix}` : prefix);
  }
}

export const logger = new DebugLogger();

export const noopLogger: Logger = {
  debug() {},
  info() {},
  warn() {},
  error() {},
  trace() {},
  child() { return noopLogger; },
};

File: vulms_dart/lib/src/utils/logger.dart
// ignore_for_file: avoid_print

/// Log level for SDK output.
enum LogLevel {
  silent,
  error,
  warn,
  info,
  debug,
  trace,
}

/// Abstract logger interface for the SDK.
///
/// Implement this to provide custom logging in your app.
abstract class VulmsLogger {
  void debug(String message);
  void info(String message);
  void warn(String message);
  void error(String message);
  void trace(String message);

  /// Create a child logger with a prefix.
  VulmsLogger child(String prefix);
}

/// Default logger implementation using `print`.
class DefaultLogger implements VulmsLogger {
  final LogLevel level;
  final String prefix;

  DefaultLogger({
    this.level = LogLevel.warn,
    this.prefix = '',
  });

  static const _priority = {
    LogLevel.silent: 0,
    LogLevel.error: 1,
    LogLevel.warn: 2,
    LogLevel.info: 3,
    LogLevel.debug: 4,
    LogLevel.trace: 5,
  };

  @override
  void debug(String message) {
    if (_priority[level]! >= _priority[LogLevel.debug]!) {
      print('[DEBUG${_pfx}] $message');
    }
  }

  @override
  void info(String message) {
    if (_priority[level]! >= _priority[LogLevel.info]!) {
      print('[INFO${_pfx}] $message');
    }
  }

  @override
  void warn(String message) {
    if (_priority[level]! >= _priority[LogLevel.warn]!) {
      print('[WARN${_pfx}] $message');
    }
  }

  @override
  void error(String message) {
    if (_priority[level]! >= _priority[LogLevel.error]!) {
      print('[ERROR${_pfx}] $message');
    }
  }

  @override
  void trace(String message) {
    if (_priority[level]! >= _priority[LogLevel.trace]!) {
      print('[TRACE${_pfx}] $message');
    }
  }

  @override
  VulmsLogger child(String childPrefix) {
    return DefaultLogger(
      level: level,
      prefix: prefix.isEmpty ? childPrefix : '$prefix:$childPrefix',
    );
  }

  String get _pfx => prefix.isEmpty ? '' : ':$prefix';
}

/// No-op logger that discards all output.
class NoopLogger implements VulmsLogger {
  const NoopLogger();

  @override
  void debug(String message) {}
  @override
  void info(String message) {}
  @override
  void warn(String message) {}
  @override
  void error(String message) {}
  @override
  void trace(String message) {}
  @override
  VulmsLogger child(String prefix) => const NoopLogger();
}
